rust/worker/src/system/executor.rs

```diff
@@ -0,0 +1,138 @@
+use futures::{Future, FutureExt, StreamExt};
+use std::{future::IntoFuture, sync::Arc};
+
+use futures::Stream;
+use tokio::{pin, select};
+
+use crate::system::ComponentContext;
+
+use super::{system::System, Component, Handler, StreamHandler};
+
+struct Inner<C, M>
+where
+    C: Component + Send + Sync + 'static,
+    M: Send + Sync + 'static,
+{
+    pub(super) channel_in: tokio::sync::broadcast::Sender<M>,
+    pub(super) cancellation_token: tokio_util::sync::CancellationToken,
+    pub(super) system_component: Arc<C>,
+    pub(super) system: System,
+}
+
+#[derive(Clone)]
+pub(super) struct ComponentExecutor<H, M>
+where
+    H: Handler<M> + Send + Sync + 'static,
+    M: Clone + Send + Sync + 'static,
+{
+    inner: Arc<Inner<H, M>>,
+    handler: Arc<H>,
+}
+
+impl<H, M> ComponentExecutor<H, M>
+where
+    H: Handler<M> + Send + Sync + 'static,
+    M: Clone + Send + Sync + 'static,
+{
+    pub(super) fn new(
+        channel_in: tokio::sync::broadcast::Sender<M>,
+        cancellation_token: tokio_util::sync::CancellationToken,
+        system_component: Arc<H>,
+        handler: Arc<H>,
+        system: System,
+    ) -> Self {
+        ComponentExecutor {
+            inner: Arc::new(Inner {
+                channel_in,
+                cancellation_token,
+                system_component,
+                system,
+            }),
+            handler,
+        }
+    }
+
+    pub(super) async fn run(&mut self, mut channel: tokio::sync::broadcast::Receiver<M>) {
+        loop {
+            select! {
+                    _ = self.inner.cancellation_token.cancelled() => {
+                        break;
+                    }
+                    message = channel.recv() => {
+                        match message {
+                            Ok(message) => {
+                                self.handler.handle(message,
+                                    &ComponentContext{
+                                        system: self.inner.system.clone(),
+                                        sender: self.inner.channel_in.clone(),
+                                        cancellation_token: self.inner.cancellation_token.clone(),
+                                        system_component: self.inner.system_component.clone(),
+                                    }
+                                ).await;
+                            }
+                            Err(_) => {
+                                // TODO: Log error
```

Comment:
Leaving todos for logging errors, when we do a tracing/log pass we can add that in